### DISCUSSION TITLE
Rotating PDF based on orientation of JSON

### DISCUSSION THREAD

**Discussion by orlandito24** [mindee/doctr#1370] [Q&A]
I was wondering if using the orientation provided by the JSON, there was a way I could orientate the PDF I'm reading into DocumenFile.from_pdf().

I'm trying to capture the coordinates of lines/words, but on rotated PDF's there are too many coordinates in play so I want the PDF to be straightened out so every line/word has two coordinates. Let me know if there is a better way to go about this. thank you!

**Comment by felixT2K** [comment]
Hi @orlandito24 👋🏼,

We are on it currently: #1336 
Additional with #1364 this approach will be much more robust for rotations between -90 and 90 degrees.
So it will be available with the next release.

After the release you can pass `assume_straight_pages=False`  `detect_orientation=True`  `straighten_pages=True` to the `ocr_predictor` instance, internally we will straighten the pages based on the page orientations in this case it will return "straight" boxes.

For the moment you can use `export_as_straight_boxes=True` if  `assume_straight_pages` is set to `False` (NOTE: this works atm only with PT: `db_resnet50_rotation` TF: `linknet_resnet18_rotation` as detection model)

For visibility: @odulcy-mindee

**Reply by orlandito24** [reply]
Hi @felixT2K,

I'm trying to understand the PT: TF: portion of your reply. I wasn't able to find db_resnet50_rotation anywhere? Would I be adding these as the reco/det on an ocr_predictor? Or would these be separate models that I would glue together?

**Reply by felixT2K** [reply]
Correct ! 
If you use PyTorch (PT) as backend:
```
from doctr.models import ocr_predictor
model = ocr_predictor('db_resnet50_rotation', 'crnn_vgg16_bn', pretrained=True, assume_straight_pages=False, export_as_straight_boxes=True)
```

**Reply by orlandito24** [reply]
@felixT2K well, this was the result haha. And there is no current way to make the boxes around the letters in the orientation of the page?

![blocks](https://github.com/mindee/doctr/assets/56779256/824c4787-983b-4b57-bb52-174dea86b496)

**Reply by felixT2K** [reply]
@orlandito24 Sure instead of `export_as_straight_boxes=True`  -> `export_as_straight_boxes=False` :)

**Reply by orlandito24** [reply]
@felixT2K WOW, the ocr_predictor you sent me is wayyyy better, than the one I've been using. I'm not sure if it was installing tensorflow or the crnn_vgg16_bn addition. What does the crnn do?

Would there be a way, given the 4 coordinate pairs, to sort the lines/words to read from top left to bottom right, as if the page was 0 degrees orientation? Trying to create a txt file replicating the document right sight up

**Reply by felixT2K** [reply]
CRNN is the text recognition model

To explain the pipe contains two models:
First one for text detection (in this case `db_resnet50_rotation` was trained on rotated samples) and the sec one to predict the "text" from the cropped boxes (default is db_resnet50 as det model and crnn_vgg16_bn as rec model).

We have btw much more models :)
See: 
https://mindee.github.io/doctr/using_doctr/using_models.html 
and
https://github.com/mindee/doctr/pull/1321/files  (this models are currently only available at main branch of the repository)

This has nothing to do with tensorflow nor Pytorch ^^

**Reply by felixT2K** [reply]
Mh it should ... does it not keeping the correct order if you for example print the result ? 
```
res = model(doc)
print(res.render())
```
?

**Reply by orlandito24** [reply]
@felixT2K You're right! Everything is working well now. Thank you so much. This has helped a lot!

**Reply by felixT2K** [reply]
Nice 👍🏼

**Reply by orlandito24** [reply]
@felixT2K There is one thing I'm realizing, not sure if it is possible but the render seems to be sorting blocks and printing the text inside. I would want more of a sort where it would print the lines sorted from top left to right down. Right now the floating PDF DOC PNG are being printed after the Digital Documents, Even thought the former is closer to the top left.

**Reply by felixT2K** [reply]
@orlandito24 I think that's tricky. Are the single words (PDF, DOC, PNG) sorted into one `block`  or each of these words into one `block` element ?

**Reply by felixT2K** [reply]
If it's the first case you could do your own sorting on the coordinates of the block elements

**Reply by orlandito24** [reply]
Hi @felixT2K, I was wondering what would the best ocr_predictor combo for files that are straight and I would want straight boxes around the text?
